plugins/destination/bigquery/client/write.go

```diff
@@ -0,0 +1,103 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/cloudquery/plugin-sdk/plugins"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	batchSize    = 1000
+	writeTimeout = 5 * time.Minute
+)
+
+type worker struct {
+	writeChan chan []interface{}
+}
+
+type item struct {
+	cols map[string]bigquery.Value
+}
+
+func (i *item) Save() (map[string]bigquery.Value, string, error) {
+	// we're not doing de-dup at the moment
+	return i.cols, bigquery.NoDedupeID, nil
+}
+
+func (c *Client) writeResource(ctx context.Context, table *schema.Table, resources <-chan []interface{}) error {
+	inserter := c.client.Dataset(c.datasetID).Table(table.Name).Inserter()
+	inserter.IgnoreUnknownValues = true
+	inserter.SkipInvalidRows = false
+	batch := make([]*item, 0)
+	for cols := range resources {
+		c.logger.Debug().Msg("Got resource")
+		saver := &item{
+			cols: make(map[string]bigquery.Value, len(table.Columns)),
+		}
+		for i := range cols {
+			if cols[i] == nil {
+				// save some bandwidth by not sending nil values
+				continue
+			}
+			saver.cols[table.Columns[i].Name] = cols[i]
+		}
+		c.logger.Debug().Interface("cols", saver.cols).Msg("got resource")
+		batch = append(batch, saver)
+		if len(batch) >= batchSize {
+			c.logger.Debug().Msg("Writing batch")
+			// we use a context with timeout here, because inserter.Put can retry indefinitely
+			// on retryable errors if not given a context timeout
+			timeoutCtx, cancel := context.WithTimeout(ctx, writeTimeout)
+			err := inserter.Put(timeoutCtx, batch)
+			if err != nil {
+				cancel()
```

Comment:
The linter forced me to add it. 😅 They say you need to ensure that cancel is called on every possible path, otherwise there could be resource leakage.

Reply:
The alternative is to use a `defer cancel()`, but then we should wrap the whole thing in a function so that we're not calling `defer` inside a for loop.